Add performance regress `test_ondemand_download_churn.py`

jbajic opened the pull request:
## Summary of changes

Add performance regress test `test_ondemand_download_churn.py`.

Closes https://github.com/neondatabase/neon/issues/7146

Tagging @problame 

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

jbajic commented on test_runner/performance/pageserver/pagebench/test_ondemand_download_churn.py:
```diff
@@ -0,0 +1,163 @@
+
+
+@pytest.mark.parametrize("duration", [30])
+@pytest.mark.parametrize("pgbench_scale", [get_scale_for_db(200)])
```
I have tried just putting 200 without `get_scale_for_db` but those test cannot pass since it takes a lot of time for tenant to becomes active and `ensure_pageserver_ready_for_benchmarking` fails

jbajic commented on test_runner/performance/pageserver/pagebench/test_ondemand_download_churn.py:
```diff
@@ -0,0 +1,163 @@
+@pytest.mark.parametrize("duration", [30])
+@pytest.mark.parametrize("pgbench_scale", [get_scale_for_db(200)])
+@pytest.mark.parametrize("io_engine", ["tokio-epoll-uring", "std-fs"])
+@pytest.mark.parametrize("concurrency_per_target", [1, 10])
```
Having 100 as parameter also causes the test to just fail without results

problame reviewed (commented):
Thanks for working on this! A couple of thoughts:
1. Latency stats are useful, and you copied the right pattern with the task-locals; but they need to be partitioned by the request type, i.e., is it an eviction or an on-demand download. Also, I'm afraid that the histogram range might be optimized for getpage-level latencies from the get-page-latest-lsn sub-benchmark, not on-demand download latencies. So, overall, I don't know whether it makes sense to have latency stats yet.
2. Instead, the idea behind on-demand download churn was to determine max throughput.
3. I don't think we the `n_tenants` and `pgbench_scale` cruft that you likely copied over from the get-page-latest-lsn sub-benchmark. Please read [the public Notion page](https://neondatabase.notion.site/Benchmarking-tokio-epoll-uring-on-demand-downloads-2024-04-15-newer-code-03c0fdc475c54492b44d9627b6e4e710) I linked from the issue; it describes the setup: a single tenant initialized with pgbench using the parameters described in that Notion doc should be sufficent to max out throughput form real S3 to an i3en.3xlarge EC2 instance.

Of course, if you don't want to spend the money on AWS, you can set up a local `minio` instance for testing.

I don't think we have good documentation for how to run the locally run the test suite with real S3.
Probably the best instructions for that are in
https://github.com/neondatabase/neon/blob/7d32af5ad5ec316761cb29d6a8141f4baf68735b/.github/workflows/build_and_test.yml#L369-L374
Feel free to augment [`test_runner/README.md`](Useful environment variables:) once you get it running locally.

Again, thanks for taking a stab at this, as described in the issue, it's a fairly self-contained task but requires some ramp-up with the test suite.

problame commented:
Please re-request review once you've made changes. I'll be unavailable until next Tuesday.

jbajic commented:
> The tests with 10, and 100 `concurrency_per_target` fail at the end but still produce results... I will try to investigate it a bit further.

I have found the issue and fixed it with the last commit. The requests are dropped at the end of the test and it is detected as an error in logs.

problame reviewed (commented):
Sorry for the late reply.

I merged from `main` and pushed some changes, mostly orthogonal to yours.

Couple of high-level remarks:

- The `ondemand_download_churn.rs` should produce absolute numbers instead of the per-second average. 
- the addition of `AWS_ENDPOINT_URL`: I guess you needed that for `minio`. At Neon, we use AWS_PROFILE for this sort of stuff. So, you'd use a custom aws profile for the minio S3 endpoint. An example for how to configure the profile is givin [here](https://aws.amazon.com/blogs/developer/new-improved-flexibility-when-configuring-endpoint-urls-with-the-aws-sdks-and-tools/). Would be nice if we can streamline your instructions in the README to just use AWS_PROFILE.

Again, sorry for the delayed reply. Let me know if you're still willing to pick up this PR.

jbajic commented:
> Sorry for the late reply.
> 
> I merged from `main` and pushed some changes, mostly orthogonal to yours.
> 
> Couple of high-level remarks:
> 
> * The `ondemand_download_churn.rs` should produce absolute numbers instead of the per-second average.
> * the addition of `AWS_ENDPOINT_URL`: I guess you needed that for `minio`. At Neon, we use AWS_PROFILE for this sort of stuff. So, you'd use a custom aws profile for the minio S3 endpoint. An example for how to configure the profile is givin [here](https://aws.amazon.com/blogs/developer/new-improved-flexibility-when-configuring-endpoint-urls-with-the-aws-sdks-and-tools/). Would be nice if we can streamline your instructions in the README to just use AWS_PROFILE.
> 
> Again, sorry for the delayed reply. Let me know if you're still willing to pick up this PR.

No issues, yes I would like to pick it up, and finish this test

problame commented:
Thanks, go ahead and re-request review from me if that's possible, otherwise reply to this thread.

jbajic commented:
> * The `ondemand_download_churn.rs` should produce absolute numbers instead of the per-second average.

I have changed the output to look like this now:
```json
{
  "downloads": 447,
  "evictions": 609,
  "runtime": {
    "nanos": 32640994,
    "secs": 30
  },
  "timeline_restarts": 1
}
```

> * the addition of `AWS_ENDPOINT_URL`: I guess you needed that for `minio`. At Neon, we use AWS_PROFILE for this sort of stuff. So, you'd use a custom aws profile for the minio S3 endpoint. An example for how to configure the profile is givin [here](https://aws.amazon.com/blogs/developer/new-improved-flexibility-when-configuring-endpoint-urls-with-the-aws-sdks-and-tools/). Would be nice if we can streamline your instructions in the README to just use AWS_PROFILE.
> 

I have also added usage of `AWS_PROFILE` to replace `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`. Using endpoint_url in profile still does not fix how the pageserver command is constructed, since `endpoint` needs to be explicitly defined for pageserver for it to work in:
```bash
./target/release/neon_local init --config=/tmp/tmphvpqctqc --pg-version 15 --pageserver-config-override=remote_storage={ bucket_name = "neon-bucket", bucket_region = "eu-central-1", prefix_in_bucket = "2eedf453-65aa-43b7-a1ed-379824d75180/test-download-churn-release-pg15-1-tokio-epoll-uring-30-/pageserver", endpoint = "http://127.0.0.1:9000" }
 --pageserver-config-override=page_cache_size=16384 --pageserver-config-override=max_file_descriptors=500000
```

problame commented:
The `endpoint` is optional: https://github.com/neondatabase/neon/blob/c18d3340b5e3c978a81c3db8b6f1e83cd9087e8a/libs/remote_storage/src/lib.rs#L557-L563


Also, this week, I merged a bunch of PRs that change how neon_local populates pageserver config; tl;dr: the config override stuff is gone, the `[[pageservers]]` section in `.neon/config` is gone, and `.neon/pageserver_$i/pageserver.toml` is only written once during `neon_local init`.

problame commented:
I'll try out your instructions and changes and report back.

problame commented:
@jbajic I took the time to write some more elaborate docs on test suite & remote storage: https://github.com/neondatabase/neon/pull/7664

Can you try the instructions in that PR? Keep in mind that you need my changes from that PR, so, I recommend you rebase this PR on top of it.

problame reviewed (commented):
Please re-request review once more when PR https://github.com/neondatabase/neon/pull/7664 has been merged and this PR has been rebased 🙏 
We're close to wrapping up!

jbajic commented:
> Please re-request review once more when PR #7664 has been merged and this PR has been rebased 🙏 We're close to wrapping up!

Thanks for the fix! I have rebased and removed redundant changes.

problame commented on test_runner/performance/pageserver/pagebench/test_ondemand_download_churn.py:
```diff
@@ -0,0 +1,175 @@
+        report=MetricReport.HIGHER_IS_BETTER,
+    )
+
+    metric = "downloads_bytes"
+    record(
+        metric,
+        metric_value=results[metric],
+        unit="byte",
+        report=MetricReport.HIGHER_IS_BETTER,
+    )
```
@bayandin are we fine with plain bytes metrics or is it better for ergonomics in the allure if we put a derived number here, e.g.

* total MiB instead of total bytes
* average MiB/s by dividing `downloads_bytes/runtime`

bayandin replied:
Hypothetically, throughput (`downloads_bytes/runtime`) is better, but we need to take not the runtime but the time we download something (or use another approach for it).
With the current implementation, `byte` is ok.

problame reviewed (commented):
Pushed some final changes from my side.

Requesting review from @bayandin before merging this.

problame commented:
@bayandin  I think this benchmark makes sense when run on EC2 against real S3.
https://github.com/neondatabase/neon/pull/7242

Should we mark it skip until we have that?

* https://github.com/neondatabase/neon/issues/6297

bayandin commented:
Merging main and doing a CI run to get benchmarks run in https://github.com/neondatabase/neon/pull/7717 (I've added a label to that PR)

jbajic commented:
Last commit for fixing the error:
```
ERROR test_runner/performance/pageserver/pagebench/test_ondemand_download_churn.py::test_download_churn[release-pg14-github-actions-selfhosted-100-std-fs-30] - AssertionError: Log errors on pageserver_1: (9559, '2024-05-13T14:09:35.888170Z  WARN request{method=DELETE path=/v1/tenant/f40d46f8ad8e8508aa35f94adfeb9048/timeline/7ee05d10dd8bd3b978e809763cb7f98d/layer/000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__0000000034CD1261-000000003567D801 request_id=7cf26d30-7520-460f-aff8-af55c710e2f8}: request was dropped before completing\n')
```

jbajic commented:
> It's done, @jbajic ! Thanks for your patience!

Thank you for the support and reviews!
